src/object/srv_obj.c

```diff
@@ -968,10 +968,17 @@ obj_singv_ec_rw_filter(daos_unit_oid_t *oid, daos_iod_t *iods, uint64_t *offs,
 	tgt_idx = oid->id_shard - start_shard;
 	for (i = 0; i < nr; i++) {
 		iod = &iods[i];
-		if (iod->iod_type != DAOS_IOD_SINGLE || (flags & ORF_EC) == 0)
+		if (iod->iod_type != DAOS_IOD_SINGLE)
 			continue;
+
+		/* Set iod_recx to NULL, since for EC object, the iod_recx
+		 * will be reused.
+		 */
+		iod->iod_recxs = NULL;
+		if ((flags & ORF_EC) == 0)
+			continue;
+
 		/* for singv EC */
-		D_ASSERT(iod->iod_recxs == NULL);
```

Comment:
the original code already ensured the server received iod_recxs is NULL for single value.
see the code in crt_proc_daos_iod_and_csum():
 251 >       if (proc_op == CRT_PROC_ENCODE || proc_op == CRT_PROC_FREE) {           
 252 >       >       if (iod->iod_type == DAOS_IOD_ARRAY && iod->iod_recxs != NULL)  
 253 >       >       >       existing_flags |= IOD_REC_EXIST;                        
 254 >       }                                                                       
 255                                                                                 
 256 >       rc = crt_proc_uint32_t(proc, &existing_flags);                          
 257 >       if (rc != 0)                                                            
 258 >       >       return -DER_HG;                                                 
 259                                                                                 
 260 >       if (proc_op == CRT_PROC_DECODE) {                                       
 261 >       >       if (existing_flags & IOD_REC_EXIST) {                           
 262 >       >       >       D_ALLOC_ARRAY(iod->iod_recxs, nr);                      
 263 >       >       >       if (iod->iod_recxs == NULL)                             
 264 >       >       >       >       D_GOTO(free, rc = -DER_NOMEM);                  
 265 >       >       }                                                               
 266 >       }                       

And this original assertion is to assert that.

On the other hand, you directly reset iod_recxs as NULL here, at code-level it seems not very safe as possibly cause mem-leak (if memory allocated in proc's decode case, then cannot free it at FREE case).

Reply:
yes, you are right.  Actually I just checked this patch is not needed after PR-3389 is landed.